t/002-upstream-tls.t

```diff
@@ -277,3 +277,441 @@ verify:1, error:0, depth:0, subject:"/C=US/ST=California/O=Kong Testing/CN=foo@e
+
+--- error_code: 502
+--- no_error_log
+[crit]
+[alert]
+[warn]
+
+
+
+=== TEST 9: setting insufficient verify depth with tls.set_upstream_ssl_verify_depth, verify failed
+--- http_config
+    lua_package_path "../lua-resty-core/lib/?.lua;lualib/?.lua;;";
+
+    server {
+        listen unix:$TEST_NGINX_HTML_DIR/nginx.sock ssl;
+        server_name   example.com;
+        ssl_certificate ../../cert/example.com.crt;
+        ssl_certificate_key ../../cert/example.com.key;
+
+        server_tokens off;
+
+        location /foo {
+            default_type 'text/plain';
+            more_clear_headers Date;
+            echo 'it works!';
+        }
+    }
+--- config
+    server_tokens off;
+
+    location /t {
+        access_by_lua_block {
+            local tls = require("resty.kong.tls")
+
+            -- verify depth 0 means only self signed cert is allowed
+            local ok, err = tls.set_upstream_ssl_verify_depth(0)
+            if not ok then
+                ngx.say("set_upstream_ssl_verify_depth failed: ", err)
+            end
+        }
+        proxy_ssl_trusted_certificate ../../cert/ca.crt;
+        proxy_ssl_verify on;
+        proxy_ssl_verify_depth 1; # default value
+        proxy_ssl_name example.com;
+        proxy_pass https://unix:$TEST_NGINX_HTML_DIR/nginx.sock:/foo;
+    }
+
+--- request
+GET /t
+--- response_body_like
+.+502 Bad Gateway.+
+
+--- error_log
+upstream SSL certificate verify error: (22:certificate chain too long)
+
+--- error_code: 502
+--- no_error_log
+[crit]
+[alert]
+[warn]
+
+
+
+=== TEST 10: setting deeper verify depth with tls.set_upstream_ssl_verify_depth, verify pass
+--- http_config
+    lua_package_path "../lua-resty-core/lib/?.lua;lualib/?.lua;;";
+
+    server {
+        listen unix:$TEST_NGINX_HTML_DIR/nginx.sock ssl;
+        server_name   example.com;
+        ssl_certificate ../../cert/example.com.crt;
+        ssl_certificate_key ../../cert/example.com.key;
+
+        server_tokens off;
+
+        location /foo {
+            default_type 'text/plain';
+            more_clear_headers Date;
+            echo 'it works!';
+        }
+    }
+--- config
+    server_tokens off;
+
+    location /t {
+        access_by_lua_block {
+            local tls = require("resty.kong.tls")
+
+            local ok, err = tls.set_upstream_ssl_verify_depth(2)
+            if not ok then
+                ngx.say("set_upstream_ssl_verify_depth failed: ", err)
+            end
+        }
+        proxy_ssl_trusted_certificate ../../cert/ca.crt;
+        proxy_ssl_verify on;
+        proxy_ssl_verify_depth 1; # default value
+        proxy_ssl_name example.com;
+        proxy_pass https://unix:$TEST_NGINX_HTML_DIR/nginx.sock:/foo;
+    }
+
+--- request
+GET /t
+--- response_body_like
+it works!
+
+--- error_log
+X509_check_host(): match
+
+--- error_code: 200
+--- no_error_log
+[error]
+[crit]
+[alert]
+
+
+
+=== TEST 11: setting insufficient trusted store with tls.set_upstream_ssl_trusted_store, verify failed
```

Comment:
This should be called "incorrect trusted store" instead? "insufficient" sounds a little strange...